### DISCUSSION TITLE
Feedback on machine that depends on multiple services

### DISCUSSION THREAD

**Discussion by asatraitis** [statelyai/xstate#3271] [General] [answered]
Hello!

I have been scratching my head how to model a machine which would request data from multiple (in parallel) services and assemble that data before shipping it.

I've come up with this: https://stately.ai/viz/4f8f0466-a748-4daf-a428-6271c2562c0d

I am just wondering if I am doing any anti patterns or if there's a clearly superior way to achieve same/similar behavior.  So far, I can see that adding more services (like middle name provider) would require adding more `waitFor...` states.

Feedback greatly appreciated!
Cheers!

**Comment by davidkpiano** [comment, marked as answer]
You can add `assign(...)` actions in the `onDone` of each compound state (region) in the parallel parent.

There's an idea that we've been thinking of though for combining "done data" from parallel regions:

```js
// idea

{
  type: 'parallel',
  states: {
    // compound regions each with final states + data
    one: { ... },
    two: { ... },
    three: { ... },
  },
  onDone: {
    actions: (_, event) => {
      // Combined data separated by region keys
      event.data; // { one: { ... }, two: { ... }, three: { ... } }
    }
  }
}
```
